src/esp/io/URDFParser.h

```diff
@@ -0,0 +1,596 @@
+   *
+   * @param model The URDF::Model datastructure to manipulate.
+   * @return Success or failure.
+   */
+  bool initTreeAndRoot(const std::shared_ptr<Model>& model) const;
+
+  /**
+   * @brief Parse URDF material info into a datastructure.
+   *
+   * @param material The Material datastructure to fill.
+   * @param config The source xml element to parse (e.g. <material>).
+   * @return Success or failure.
+   */
+  bool parseMaterial(Material& material, tinyxml2::XMLElement* config) const;
+
+  /**
+   * @brief Parse joint limits from a <joint> xml element into a datastructure.
+   *
+   * @param joint The Joint datastructure to fill.
+   * @param config The source xml element to parse (e.g. <joint>).
+   * @return Success or failure.
+   */
+  bool parseJointLimits(Joint& joint, tinyxml2::XMLElement* config) const;
+
+  /**
+   * @brief Parse joint dynamic info from a <joint> xml element into a
+   * datastructure.
+   *
+   * @param joint The Joint datastructure to fill.
+   * @param config The source xml element to parse (e.g. <joint>).
+   * @return Success or failure.
+   */
+  bool parseJointDynamics(Joint& joint, tinyxml2::XMLElement* config) const;
+
+  /**
+   * @brief Parse joint info from a <joint> xml element into a datastructure.
+   *
+   * @param joint The Joint datastructure to fill.
+   * @param config The source xml element to parse (e.g. <joint>).
+   * @return Success or failure.
+   */
+  bool parseJoint(Joint& joint, tinyxml2::XMLElement* config);
+
+  /**
+   * @brief Parse link info from a <link> xml element into a datastructure.
+   *
+   * @param link The Link datastructure to fill.
+   * @param config The source xml element to parse (e.g. <link>).
+   * @return Success or failure.
+   */
+  bool parseLink(const std::shared_ptr<Model>&,
+                 Link& link,
+                 tinyxml2::XMLElement* config);
+
+  /**
+   * @brief Parse sensor info from a into a datastructure.
+   *
+   * TODO: Not implemented yet, does nothing.
+   *
+   * @param model The URDF::Model datastructure to fill.
+   * @param link The Link datastructure to fill.
+   * @param joint The Joint datastructure to fill.
+   * @param config The source xml element to parse.
+   * @return Success or failure.
+   */
+  bool parseSensor(CORRADE_UNUSED const std::shared_ptr<Model>& model,
+                   CORRADE_UNUSED Link& link,
+                   CORRADE_UNUSED Joint& joint,
+                   CORRADE_UNUSED tinyxml2::XMLElement* config) {
+    // TODO: this
+    return false;
+  };
+
+  /**
+   * @brief Check whether or not a given filepath exists.
+   *
+   * @param filename The filepath to check.
+   * @return Filepath is valid or invalid.
+   */
+  bool validateMeshFile(std::string& filename);
+
+ public:
+  Parser() = default;
+
+  // parse a loaded URDF string into relevant general data structures
+  // return false if the string is not a valid urdf or other error causes abort
+  bool parseURDF(std::shared_ptr<Model>& model,
```

Comment:
Nit: kinda a strange interface to pass in a reference to a shared_ptr to be populated. How about just return a valid shared_ptr or null on failure?